chia/data_layer/data_store.py

```diff
@@ -821,3 +879,223 @@ async def get_operations(
+                                        lock=False,
+                                    )
+                                else:
+                                    await self.delete(
+                                        bytes.fromhex(row["key"]),
+                                        tree_id,
+                                        status=Status(row["root_status"]),
+                                        lock=False,
+                                    )
+                                current_root = await self.get_tree_root(tree_id, lock=False)
+                                if current_root.node_hash is None:
+                                    if row["hash"] != "None":
+                                        return False
+                                else:
+                                    if current_root.node_hash.hex() != row["hash"]:
+                                        return False
+                                existing_generation += 1
+
+                        await ws.send_str(json.dumps({"type": "close"}))
+
+                return True
+
+            return False
+
+    async def subscribe(self, subscription: Subscription, *, lock: bool = True) -> None:
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            await self.db.execute(
+                "INSERT INTO subscriptions(tree_id, mode, ip, port, wallet_generation) "
+                "VALUES (:tree_id, :mode, :ip, :port, 0)",
+                {
+                    "tree_id": subscription.tree_id.hex(),
+                    "mode": subscription.mode.value,
+                    "ip": subscription.ip,
+                    "port": subscription.port,
+                },
+            )
+
+    async def unsubscribe(self, tree_id: bytes32, *, lock: bool = True) -> None:
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            await self.db.execute(
+                "DELETE FROM subscriptions WHERE tree_id == :tree_id",
+                {"tree_id": tree_id.hex()},
+            )
+            await self.db.execute(
+                "DELETE FROM root WHERE tree_id == :tree_id",
+                {"tree_id": tree_id.hex()},
+            )
+
+    async def get_subscriptions(self, *, lock: bool = True) -> List[Subscription]:
+        subscriptions: List[Subscription] = []
+
+        async with self.db_wrapper.locked_transaction(lock=lock):
+            cursor = await self.db.execute(
+                "SELECT * from subscriptions",
+            )
+            async for row in cursor:
+                tree_id = bytes32.fromhex(row["tree_id"])
+                mode_value = int(row["mode"])
+                ip = row["ip"]
+                port = uint16(row["port"])
+                subscriptions.append(Subscription(tree_id, DownloadMode(mode_value), ip, port))
+
+        return subscriptions
+
+    async def get_wallet_generation(self, tree_id: bytes32, *, lock: bool = True) -> int:
```

Comment:
this should be done by the wallet, the wallet already manages history for the singleton

Reply:
this is just to track how much we've validated before - it compares against `singleton_record.generation`, which is the target, not the value produced by this.